internal/incident/incident.go

```diff
@@ -354,20 +368,50 @@ func (i *Incident) evaluateEscalations() {
 				if escalation.Condition == nil {
 					matched = true
 				} else {
-					cond := &rule.EscalationFilter{
-						IncidentAge:      time.Now().Sub(i.StartedAt),
-						IncidentSeverity: i.Severity(),
+					evaluateCond := func(cond *rule.EscalationFilter) (bool, error) {
+						before := &rule.EscalationFilter{
+							IncidentAge:      cond.IncidentAge - time.Nanosecond,
+							IncidentSeverity: cond.IncidentSeverity,
+						}
+						matched, err := escalation.Condition.Eval(before)
+						if err != nil {
+							return false, err
+						}
+
+						if !matched {
+							matched, err = escalation.Condition.Eval(cond)
+							if err != nil {
+								return false, err
+							}
+						}
+
+						if !matched {
+							return escalation.Condition.Eval(
+								&rule.EscalationFilter{IncidentAge: cond.IncidentAge + time.Nanosecond, IncidentSeverity: cond.IncidentSeverity},
+							)
+						}
+
+						return matched, nil
 					}
 
 					var err error
-					matched, err = escalation.Condition.Eval(cond)
+					matched, err = evaluateCond(cond)
 					if err != nil {
 						i.logger.Warnw(
 							"Failed to evaluate escalation condition", zap.String("rule", r.Name),
 							zap.String("escalation", escalation.DisplayName()), zap.Error(err),
 						)
 
 						matched = false
+					} else if !matched {
+						age := cond.ExtractRetryAfter(escalation.Condition.ExtractConditions(), cond.IncidentAge)
+						if age > 0 && (retryAfter <= 0 || retryAfter > age) {
+							retryAfter = time.Until(i.StartedAt.Add(age))
+							i.scheduler.condition = cond
+							i.scheduler.condition.IncidentAge = age
+						} else {
+							i.scheduler.condition = nil
+						}
```

Comment:
Please explain what the value in `i.scheduler.condition` is supposed to express. I find it hard to follow what's going on here, `condition` stores an `EscalationFilter` which actually is more like a filter context (i.e. stores the context in which a filter is evaluated), but now it stores the time after which the filter is supposed to be reevaluated?

Reply:
`i.scheduler.condition` just stores the current escalation filter condition from which the `retryAfter` is extracted and modifies it's `IncidentAge` threshold and never stores `the time after which the filter is supposed to be reevaluated?`. This is needed because when you construct the escalation filter each time evaluating the escalations like:
https://github.com/Icinga/icinga-notifications/blob/14f00144b0fa82d5e40e70d3dbd70bb4588ff017/internal/incident/incident.go#L348

The condition won't match even it's retried after the correct **extracted** retry after interval.

Reply:
Please try to explain what these values are supposed to represent, without just explaining how they're set/accessed. So for example for the `Rules` member of the struct, that explanation would be something like "a set of all rules that are active on this incident and influence how notifications are sent" (instead of something like "rules are added when they match on this incident").

Reply:
After some more looking at this, I think you're trying to store the `EscalationFilter` context for the time in the future when the escalation conditions are due to be evaluated the next time. However, and that's probably what has irritated me, you don't seem to consider that other external events can happen in that time. So if you have an escalation for `incident_age>=2h` and after 1h, the incident goes from warning to crit, that has to affect the context that's used for evaluating the condition, however you seem to always use the already stored one.